### PR TITLE
Filters for periodic design patterns

### PR DESCRIPTION
**Pull request by mawc2019**
This PR makes the 2d filters in `filters.py` applicable to periodic 2d designs, where the design pattern is periodic in _x_, _y_, or both _x_ and _y_ directions. In `cylindrical_filter`, `conic_filter`, and `gaussian_filter`, and the periodicity is specified by a new argument `periodic_axes`, whose value can be `0`, `1`, `(0,1)`, or `None`.

In most cases where the filter diameter is not larger than the size of the input design pattern, extra padding along the periodic axis is simply not added before convolution. If the filter diameter is larger than the size of the input design pattern along a periodic axis, before convolution, the design pattern is duplicated along that axis so that the resultant design pattern along that axis is not smaller than the filter diameter.

The change in those filter functions also makes `constraint_solid` and `constraint_void` applicable to periodic 2d designs.

By the way, the [missing `+1`](https://github.com/NanoComp/meep/discussions/2240) in calculating the array size for the material grid is restored in this PR.

### CODE REVIEW

**Comment by mawc2019** [comment]
In this example, the function `conic_filter` with `radius = 55` and `resolution = 1` is used.
image

**Comment by mawc2019** [comment]
After restoring `+1` in calculating the array size for the material grid, some tests in [`test_adjoint_solver.py`](https://txe1-portal.mit.edu/fw2/d-5-3-4-12792/notebooks/repo/meep/python/tests/test_adjoint_solver.py) fail when the single-precision floating point is used. The relevant tolerances need to be increased in order to make those tests pass.

Python 3.7:
```
======================================================================
FAIL: test_eigenmode (__main__.TestAdjointSolver)
Verifies that the adjoint gradient for an objective function based
----------------------------------------------------------------------
Traceback (most recent call last):
  File "../../../python/tests/test_adjoint_solver.py", line 592, in test_eigenmode
    self.assertClose(adj_dd, fnd_dd, epsilon=tol)
  File "/home/runner/work/meep/meep/build/meep-1.27.0-beta/python/tests/utils.py", line 30, in assertClose
    self.assertLessEqual(diff_norm, epsilon * np.maximum(x_norm, y_norm), msg)
AssertionError: 2.4692787936065122e-05 not less than or equal to 2.147141650190434e-05 : 

======================================================================
FAIL: test_ldos (__main__.TestAdjointSolver)
Verifies that the adjoint gradient for an objective function based
----------------------------------------------------------------------
Traceback (most recent call last):
  File "../../../python/tests/test_adjoint_solver.py", line 621, in test_ldos
    self.assertClose(adj_dd, fnd_dd, epsilon=tol)
  File "/home/runner/work/meep/meep/build/meep-1.27.0-beta/python/tests/utils.py", line 30, in assertClose
    self.assertLessEqual(diff_norm, epsilon * np.maximum(x_norm, y_norm), msg)
AssertionError: 1.866014771444676e-10 not less than or equal to 7.504694841953264e-11 : 
```
Python 3.10:
```
======================================================================

FAIL: test_DFT_fields (__main__.TestAdjointSolver)
Verifies that the adjoint gradient for an objective function based
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/meep/meep/build/meep-1.27.0-beta/_build/sub/python/../../../python/tests/test_adjoint_solver.py", line 551, in test_DFT_fields
    self.assertClose(adj_dd, fnd_dd, epsilon=tol)
  File "/home/runner/work/meep/meep/build/meep-1.27.0-beta/python/tests/utils.py", line 30, in assertClose
    self.assertLessEqual(diff_norm, epsilon * np.maximum(x_norm, y_norm), msg)
AssertionError: 0.00032576555598241895 not less than or equal to 0.00015930181276331502 : 

======================================================================
FAIL: test_eigenmode (__main__.TestAdjointSolver)
Verifies that the adjoint gradient for an objective function basedFAILED (failures=3)

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/meep/meep/build/meep-1.27.0-beta/_build/sub/python/../../../python/tests/test_adjoint_solver.py", line 592, in test_eigenmode
    self.assertClose(adj_dd, fnd_dd, epsilon=tol)
  File "/home/runner/work/meep/meep/build/meep-1.27.0-beta/python/tests/utils.py", line 30, in assertClose
    self.assertLessEqual(diff_norm, epsilon * np.maximum(x_norm, y_norm), msg)
AssertionError: 6.287598863467803e-05 not less than or equal to 3.214987858228824e-05 : 

======================================================================
FAIL: test_ldos (__main__.TestAdjointSolver)
Verifies that the adjoint gradient for an objective function based
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/meep/meep/build/meep-1.27.0-beta/_build/sub/python/../../../python/tests/test_adjoint_solver.py", line 621, in test_ldos
    self.assertClose(adj_dd, fnd_dd, epsilon=tol)
  File "/home/runner/work/meep/meep/build/meep-1.27.0-beta/python/tests/utils.py", line 30, in assertClose
    self.assertLessEqual(diff_norm, epsilon * np.maximum(x_norm, y_norm), msg)
AssertionError: 2.0344705001442236e-10 not less than or equal to 1.5012644966061473e-10 : 
```

**Comment by stevengj** [comment]
Just split the off-by-one changes to a separate PR.

File: python/adjoint/filters.py [hunk 1]
```diff
@@ -90,23 +101,44 @@ def simple_2d_filter(x, h):
         Input array to be filtered. Must be 2D.
     h : array_like (2D)
         Filter kernel (before the DFT). Must be same size as `x`
+    periodic_axes: array_like (1D)
+        Array that indicates the input array `x` is periodic along which axes.
```
Comment (stevengj) [inline python/adjoint/filters.py:105]:
```suggestion
        List of axes (0, 1, and or 2) that are to be treated as periodic (defaulting to none).
```

**Comment by mawc2019** [comment]
This PR also makes lengthscale constraint functions applicable to 2d periodic design patterns. The new test in `test_adjoint_solver.py` checks that outputs of `constraint_solid()` and `constraint_void()` are invariant when a design pattern is shifted along periodic directions. Because each of the lengthscale constraint functions contains a filter function, tests exclusively for periodic filter functions are not included.

File: python/tests/test_adjoint_solver.py [hunk 2]
```diff
@@ -943,6 +943,123 @@ def test_mode_source_bandwidth(self):
...
+        idx_row_shift, idx_col_shift = int(0.19 * self.Nx), int(0.28 * self.Ny)
+        p = self.p.reshape(self.Nx, self.Ny)
+        p_row_shift = np.vstack((p[idx_row_shift:, :], p[0:idx_row_shift, :]))
+        p_col_shift = np.hstack((p[:, idx_col_shift:], p[:, 0:idx_col_shift]))
+
+        eta_e = 0.55
+        eta_d = 1 - eta_e
+        beta, eta = 10, 0.5
+        radius, c = 0.3, 400
+        places = 18
```
Comment (oskooi) [inline python/tests/test_adjoint_solver.py:961]: This seems way too low (smaller than machine precision). Are you sure this test is actually doing anything with the tolerance set so low?

Comment (stevengj) [reply python/tests/test_adjoint_solver.py:961]: By making the tolerance so low it's basically checking that the value is invariant to machine precision.

Comment (stevengj) [reply python/tests/test_adjoint_solver.py:961]: But I agree that the `places` should probably be smaller.

Comment (stevengj) [reply python/tests/test_adjoint_solver.py:961]: @mawc2019 says that the absolute magnitude of the constraint functions is about `1e-5`, so this is asking it to match to 13 digits.   So this is fine.

**Pull request merged**